models/src/main/protobuf/CasperMessage.proto

```diff
@@ -180,24 +180,24 @@ message BlockMessage {
 message Header {
   repeated bytes parentsHashList = 1; //list of parent block hashes
   bytes postStateHash      = 2;
-  bytes newCodeHash        = 3;
-  bytes commReductionsHash = 4;
+  bytes deploysHash        = 3;
   int64 timestamp = 5;
   int64 version = 6;
   int32 deployCount = 7;
   bytes extraBytes = 8;
 }
 
-message DeployCost {
-    Deploy deploy = 1; //Sha256 hash of the DeployData
-    PCost Cost = 2 ;
+message ProcessedDeploy {
+    Deploy deploy = 1;
+    PCost cost = 2 ;
+    repeated Event log = 3; //the new terms and comm. rule reductions from this deploy
+    bool errored = 4; //true if deploy encountered a user error
 }
```

Comment:
It is because I wanted to make deploys atomic; i.e. they fully complete or they have no effect on the post-state. But you need to replay all deploys, including the ones that ended in error, in order to verify that the results in the block are correct. This means that in the case of deploy errors you need to replay its log and then revert only those changes. Storing the log with each deploy clearly delineates which effects should not be persisted due to errors. 

 However, I suppose what I could do is instead keep all the error-free deploys together, with a single common log and only have the separate logs for the deploys with errors. But this seems more complicated and introduces some ambiguity around when to replay which deploys (because the order can matter).

Reply:
Since the erroring deploys get rolled back and you have no guarantees when the block creator was going to run your deploy in the first place, it is fine to just run all the erroring deploys last. If any of the claimed to be erroring deploys is actually not erroring, then you can slash.

Reply:
Oh I see. Yeah we can keep around a error log coupled with each deploy and have it be an empty Seq for valid deploys.

Reply:
Which makes me think we should have repeated Processed Deploys and repeated ErroredDeploys instead of having a boolean flag in processed deploys.

Reply:
I think this is a good change :+1: Having events coupled with deploy that caused them.